fix: Update watsonx.ai provider to use LiteLLM mixin and list all models

jwm4 opened the pull request:
# What does this PR do?

- The watsonx.ai provider now uses the LiteLLM mixin instead of using IBM's library, which does not seem to be working (see #3165 for context).
- The watsonx.ai provider now lists all the models available by calling the watsonx.ai server instead of having a hard coded list of known models.  (That list gets out of date quickly)
- An edge case in [llama_stack/core/routers/inference.py](https://github.com/llamastack/llama-stack/pull/3674/files#diff-a34bc966ed9befd9f13d4883c23705dff49be0ad6211c850438cdda6113f3455) is addressed that was causing my manual tests to fail.
- Fixes `b64_encode_openai_embeddings_response` which was trying to enumerate over a dictionary and then reference elements of the dictionary using .field instead of ["field"].  That method is called by the LiteLLM mixin for embedding models, so it is needed to get the watsonx.ai embedding models to work.
- A unit test along the lines of the one in #3348 is added.  A more comprehensive plan for automatically testing the end-to-end functionality for inference providers would be a good idea, but is out of scope for this PR.
- Updates to the watsonx distribution.  Some were in response to the switch to LiteLLM (e.g., updating the Python packages needed).  Others seem to be things that were already broken that I found along the way (e.g., a reference to a watsonx specific doc template that doesn't seem to exist).

Closes #3165

Also it is related to a line-item in #3387 but doesn't really address that goal (because it uses the LiteLLM mixin, not the OpenAI one).  I tried the OpenAI one and it doesn't work with watsonx.ai, presumably because the watsonx.ai service is not OpenAI compatible.  It works with LiteLLM because LiteLLM has a provider implementation for watsonx.ai.

## Test Plan

The test script below goes back and forth between the OpenAI and watsonx providers.  The idea is that the OpenAI provider shows how it should work and then the watsonx provider output shows that it is also working with watsonx. Note that the result from the MCP test is not as good (the Llama 3.3 70b model does not choose tools as wisely as gpt-4o), but it is still working and providing a valid response.  For more details on setup and the MCP server being used for testing, see [the AI Alliance sample notebook](https://github.com/The-AI-Alliance/llama-stack-examples/blob/main/notebooks/01-responses/) that these examples are drawn from.

```python
#!/usr/bin/env python3

import json
from llama_stack_client import LlamaStackClient
from litellm import completion
import http.client


def print_response(response):
    """Print response in a nicely formatted way"""
    print(f"ID: {response.id}")
    print(f"Status: {response.status}")
    print(f"Model: {response.model}")
    print(f"Created at: {response.created_at}")
    print(f"Output items: {len(response.output)}")
    
    for i, output_item in enumerate(response.output):
        if len(response.output) > 1:
            print(f"\n--- Output Item {i+1} ---")
        print(f"Output type: {output_item.type}")
        
        if output_item.type in ("text", "message"):
            print(f"Response content: {output_item.content[0].text}")
        elif output_item.type == "file_search_call":
            print(f"  Tool Call ID: {output_item.id}")
            print(f"  Tool Status: {output_item.status}")
            # 'queries' is a list, so we join it for clean printing
            print(f"  Queries: {', '.join(output_item.queries)}")
            # Display results if they exist, otherwise note they are empty
            print(f"  Results: {output_item.results if output_item.results else 'None'}")
        elif output_item.type == "mcp_list_tools":
            print_mcp_list_tools(output_item)
        elif output_item.type == "mcp_call":
            print_mcp_call(output_item)
        else:
            print(f"Response content: {output_item.content}")


def print_mcp_call(mcp_call):
    """Print MCP call in a nicely formatted way"""
    print(f"\n🛠️  MCP Tool Call: {mcp_call.name}")
    print(f"   Server: {mcp_call.server_label}")
    print(f"   ID: {mcp_call.id}")
    print(f"   Arguments: {mcp_call.arguments}")
    
    if mcp_call.error:
        print("Error: {mcp_call.error}")
    elif mcp_call.output:
        print("Output:")
        # Try to format JSON output nicely
        try:
            parsed_output = json.loads(mcp_call.output)
            print(json.dumps(parsed_output, indent=4))
        except:
            # If not valid JSON, print as-is
            print(f"   {mcp_call.output}")
    else:
        print("   ⏳ No output yet")


def print_mcp_list_tools(mcp_list_tools):
    """Print MCP list tools in a nicely formatted way"""
    print(f"\n🔧 MCP Server: {mcp_list_tools.server_label}")
    print(f"   ID: {mcp_list_tools.id}")
    print(f"   Available Tools: {len(mcp_list_tools.tools)}")
    print("=" * 80)
    
    for i, tool in enumerate(mcp_list_tools.tools, 1):
        print(f"\n{i}. {tool.name}")
        print(f"   Description: {tool.description}")
        
        # Parse and display input schema
        schema = tool.input_schema
        if schema and 'properties' in schema:
            properties = schema['properties']
            required = schema.get('required', [])
            
            print("   Parameters:")
            for param_name, param_info in properties.items():
                param_type = param_info.get('type', 'unknown')
                param_desc = param_info.get('description', 'No description')
                required_marker = " (required)" if param_name in required else " (optional)"
                print(f"     • {param_name} ({param_type}){required_marker}")
                if param_desc:
                    print(f"       {param_desc}")
        
        if i < len(mcp_list_tools.tools):
            print("-" * 40)


def main():
    """Main function to run all the tests"""
    
    # Configuration
    LLAMA_STACK_URL = "http://localhost:8321/"
    LLAMA_STACK_MODEL_IDS = [
        "openai/gpt-3.5-turbo",
        "openai/gpt-4o",
        "llama-openai-compat/Llama-3.3-70B-Instruct",
        "watsonx/meta-llama/llama-3-3-70b-instruct"
    ]
    
    # Using gpt-4o for this demo, but feel free to try one of the others or add more to run.yaml.
    OPENAI_MODEL_ID = LLAMA_STACK_MODEL_IDS[1]
    WATSONX_MODEL_ID = LLAMA_STACK_MODEL_IDS[-1]
    NPS_MCP_URL = "http://localhost:3005/sse/"
    
    print("=== Llama Stack Testing Script ===")
    print(f"Using OpenAI model: {OPENAI_MODEL_ID}")
    print(f"Using WatsonX model: {WATSONX_MODEL_ID}")
    print(f"MCP URL: {NPS_MCP_URL}")
    print()
    
    # Initialize client
    print("Initializing LlamaStackClient...")
    client = LlamaStackClient(base_url="http://localhost:8321")
    
    # Test 1: List models
    print("\n=== Test 1: List Models ===")
    try:
        models = client.models.list()
        print(f"Found {len(models)} models")
    except Exception as e:
        print(f"Error listing models: {e}")
        raise e
    
    # Test 2: Basic chat completion with OpenAI
    print("\n=== Test 2: Basic Chat Completion (OpenAI) ===")
    try:
        chat_completion_response = client.chat.completions.create(
            model=OPENAI_MODEL_ID,
            messages=[{"role": "user", "content": "What is the capital of France?"}]
        )
        
        print("OpenAI Response:")
        for chunk in chat_completion_response.choices[0].message.content:
            print(chunk, end="", flush=True)
        print()
    except Exception as e:
        print(f"Error with OpenAI chat completion: {e}")
        raise e
    
    # Test 3: Basic chat completion with WatsonX
    print("\n=== Test 3: Basic Chat Completion (WatsonX) ===")
    try:
        chat_completion_response_wxai = client.chat.completions.create(
            model=WATSONX_MODEL_ID,
            messages=[{"role": "user", "content": "What is the capital of France?"}],
        )
        
        print("WatsonX Response:")
        for chunk in chat_completion_response_wxai.choices[0].message.content:
            print(chunk, end="", flush=True)
        print()
    except Exception as e:
        print(f"Error with WatsonX chat completion: {e}")
        raise e
    
    # Test 4: Tool calling with OpenAI
    print("\n=== Test 4: Tool Calling (OpenAI) ===")
    tools = [
        {
            "type": "function",
            "function": {
                "name": "get_current_weather",
                "description": "Get the current weather for a specific location",
                "parameters": {
                    "type": "object",
                    "properties": {
                        "location": {
                            "type": "string",
                            "description": "The city and state, e.g., San Francisco, CA",
                        },
                        "unit": {
                            "type": "string",
                            "enum": ["celsius", "fahrenheit"]
                        },
                    },
                    "required": ["location"],
                },
            },
        }
    ]
    
    messages = [
        {"role": "user", "content": "What's the weather like in Boston, MA?"}
    ]
    
    try:
        print("--- Initial API Call ---")
        response = client.chat.completions.create(
            model=OPENAI_MODEL_ID,
            messages=messages,
            tools=tools,
            tool_choice="auto",  # "auto" is the default
        )
        print("OpenAI tool calling response received")
    except Exception as e:
        print(f"Error with OpenAI tool calling: {e}")
        raise e
    
    # Test 5: Tool calling with WatsonX
    print("\n=== Test 5: Tool Calling (WatsonX) ===")
    try:
        wxai_response = client.chat.completions.create(
            model=WATSONX_MODEL_ID,
            messages=messages,
            tools=tools,
            tool_choice="auto",  # "auto" is the default
        )
        print("WatsonX tool calling response received")
    except Exception as e:
        print(f"Error with WatsonX tool calling: {e}")
        raise e
    
    # Test 6: Streaming with WatsonX
    print("\n=== Test 6: Streaming Response (WatsonX) ===")
    try:
        chat_completion_response_wxai_stream = client.chat.completions.create(
            model=WATSONX_MODEL_ID,
            messages=[{"role": "user", "content": "What is the capital of France?"}],
            stream=True
        )
        print("Model response: ", end="")
        for chunk in chat_completion_response_wxai_stream:
            # Each 'chunk' is a ChatCompletionChunk object.
            # We want the content from the 'delta' attribute.
            if hasattr(chunk, 'choices') and chunk.choices is not None:
                content = chunk.choices[0].delta.content
                # The first few chunks might have None content, so we check for it.
                if content is not None:
                    print(content, end="", flush=True)
        print()
    except Exception as e:
        print(f"Error with streaming: {e}")
        raise e
    
    # Test 7: MCP with OpenAI
    print("\n=== Test 7: MCP Integration (OpenAI) ===")
    try:
        mcp_llama_stack_client_response = client.responses.create(
            model=OPENAI_MODEL_ID,
            input="Tell me about some parks in Rhode Island, and let me know if there are any upcoming events at them.",
            tools=[
                {
                    "type": "mcp",
                    "server_url": NPS_MCP_URL,
                    "server_label": "National Parks Service tools",
                    "allowed_tools": ["search_parks", "get_park_events"],
                }
            ]
        )
        print_response(mcp_llama_stack_client_response)
    except Exception as e:
        print(f"Error with MCP (OpenAI): {e}")
        raise e
    
    # Test 8: MCP with WatsonX
    print("\n=== Test 8: MCP Integration (WatsonX) ===")
    try:
        mcp_llama_stack_client_response = client.responses.create(
            model=WATSONX_MODEL_ID,
            input="What is the capital of France?"
        )
        print_response(mcp_llama_stack_client_response)
    except Exception as e:
        print(f"Error with MCP (WatsonX): {e}")
        raise e
    
    # Test 9: MCP with Llama 3.3
    print("\n=== Test 9: MCP Integration (Llama 3.3) ===")
    try:
        mcp_llama_stack_client_response = client.responses.create(
            model=WATSONX_MODEL_ID,
            input="Tell me about some parks in Rhode Island, and let me know if there are any upcoming events at them.",
            tools=[
                {
                    "type": "mcp",
                    "server_url": NPS_MCP_URL,
                    "server_label": "National Parks Service tools",
                    "allowed_tools": ["search_parks", "get_park_events"],
                }
            ]
        )
        print_response(mcp_llama_stack_client_response)
    except Exception as e:
        print(f"Error with MCP (Llama 3.3): {e}")
        raise e
    
    # Test 10: Embeddings
    print("\n=== Test 10: Embeddings ===")
    try:
        conn = http.client.HTTPConnection("localhost:8321")
        payload = json.dumps({
            "model": "watsonx/ibm/granite-embedding-278m-multilingual",
            "input": "Hello, world!",
        })
        headers = {
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        }
        conn.request("POST", "/v1/openai/v1/embeddings", payload, headers)
        res = conn.getresponse()
        data = res.read()
        print(data.decode("utf-8"))
    except Exception as e:
        print(f"Error with Embeddings: {e}")
        raise e

    print("\n=== Testing Complete ===")


if __name__ == "__main__":
    main()
```

leseb commented on llama_stack/providers/remote/inference/watsonx/watsonx.py:
```diff
@@ -4,246 +4,115 @@
+        self.available_models = None
+        self.config = config
+
+    # get_api_key = LiteLLMOpenAIMixin.get_api_key
```
leftover?

jwm4 replied:
Yes, will remove in the next version, once I am able to get local pre-commits working.

leseb commented on llama_stack/providers/remote/inference/watsonx/watsonx.py:
```diff
@@ -4,246 +4,115 @@
+        """
+        Retrieves foundation model specifications from the watsonx.ai API.
+        """
+        url = f"{self.config.url}/ml/v1/foundation_model_specs?version=2023-10-25"
```
```suggestion
        url = f"{self.config.url}/ml/v1/foundation_model_specs?version=2025-09-03"
```

Can we use a more recent version to get more models? We might want to expose a config flag for this?

jwm4 replied:
I meant to put in a comment here explaining why this is hard-coded, but I forgot.  I will include that comment in the next update and move to the latest API date version (2024-03-14).  In general we don't want to have a config flag for API date versions, because they only change when there is a breaking change to the API, so we shouldn't assume the code will work as implemented with a future API date version.  In any event, this API returns all the currently available models regardless of what API date version is used.

mattf commented on llama_stack/core/routers/inference.py:
```diff
@@ -611,7 +611,7 @@ async def stream_tokens_and_compute_metrics_openai_chat(
 
                     # Add metrics to the chunk
-                    if self.telemetry and chunk.usage:
+                    if self.telemetry and hasattr(chunk, "usage") and chunk.usage:
```
this looks unrelated

mattf replied:
does watsonx return usage information? we need to fix/adapt in the adapter, not in the core. putting this in the core obscures the issue.

mattf commented on llama_stack/providers/remote/inference/watsonx/watsonx.py:
```diff
@@ -4,246 +4,114 @@
+
+class WatsonXInferenceAdapter(LiteLLMOpenAIMixin):
+    _config: WatsonXConfig
+    __provider_id__: str = "watsonx"
```
this gets set by the server, see https://github.com/llamastack/llama-stack/blob/main/llama_stack/core/resolver.py#L397

mattf commented on llama_stack/providers/remote/inference/watsonx/watsonx.py:
```diff
@@ -4,246 +4,114 @@
-                    seen_finish_reason = True
-                    break
+        self.available_models = None
+        self.config = config
```
you called it _config above

mattf commented on llama_stack/providers/remote/inference/watsonx/watsonx.py:
```diff
@@ -4,246 +4,114 @@
+    def get_base_url(self) -> str:
+        return self.config.url
+
+    async def initialize(self):
```
no need to override - https://github.com/llamastack/llama-stack/blob/main/llama_stack/providers/utils/inference/litellm_openai_mixin.py#L82

mattf commented on llama_stack/providers/remote/inference/watsonx/watsonx.py:
```diff
@@ -4,246 +4,114 @@
+    async def initialize(self):
+        await super().initialize()
+
+    async def shutdown(self):
```
no need to override - https://github.com/llamastack/llama-stack/blob/main/llama_stack/providers/utils/inference/litellm_openai_mixin.py#L85

mattf commented on llama_stack/providers/utils/inference/openai_compat.py:
```diff
@@ -1414,12 +1414,12 @@ def b64_encode_openai_embeddings_response(
         if encoding_format == "base64":
             byte_array = bytearray()
-            for embedding_value in embedding_data.embedding:
+            for embedding_value in embedding_data["embedding"]:
```
careful here. we have lots of paths to these helpers, some pass types others pass dicts. if you can, pass a type.

i agree, the signature hints are probably wrong

mattf replied:
that's a good idea

mattf commented on llama_stack/providers/utils/inference/litellm_openai_mixin.py:
```diff
@@ -334,6 +336,7 @@ async def openai_chat_completion(
             api_key=self.get_api_key(),
             api_base=self.api_base,
         )
+        logger.info(f"params to litellm (openai compat): {params}")
```
avoid this, it'll not only produce a lot of output but it'll also log user data

jwm4 replied:
Good catch, thx!  I put that in there thinking I wasn't pushing this file anyway, and then forgot to take it out when we wound up moving the b64_encode_openai_embeddings_response method in to the class.  Fixed now.

mattf reviewed (approved):
lgtm

mattf commented:
as a bonus, if you can record inference responses for watsonx we'll be able to make additional changes w/o needing a key

mattf commented on llama_stack/providers/remote/inference/watsonx/watsonx.py:
```diff
@@ -4,246 +4,103 @@
-        )
-        return await self._get_openai_client().completions.create(**params)  # type: ignore
+    async def check_model_availability(self, model):
+        return True
```
why is this always true?

jwm4 replied:
I thought that was the way to signal that we don't have a static list, but I looked more closely at how the OpenAI mixin handles this issue, and I reimplemented to mirror what it does.

jwm4 commented on tests/unit/providers/inference/test_inference_client_caching.py:
```diff
@@ -65,3 +60,29 @@ def test_openai_provider_data_used(config_cls, adapter_cls, provider_data_valida
             assert inference_adapter.client.api_key == api_key
+
+
+@pytest.mark.parametrize(
```
The goal of this function is to be a general-purposes test for whether the x-llamastack-provider-data API key header for inference providers that use the LiteLLM Mixin just like the one above tests the inference providers that use the OpenAI Mixin.  Right now we only have one LIteLLM-based provider, but having a list here makes it easier to add more of these tests if/when we add more LIteLLM-based providers which I think we would do if there was a big demand for another LLM service that isn't OpenAI compatible.

leseb commented:
@jwm4, the issue you referenced was actually fixed a few weeks ago. So, switching from ibm_watsonx_ai to litellm is not the fix. Help me understand why we need to switch? Thanks!

jwm4 commented:
@leseb here is what happened:
- On the community call, @mattf announced that he was looking to get all the providers to use the mixins and asked for volunteers for the ones that were missing.  The idea was that we didn't want a bunch of bespoke logic and dependencies for each provider.
- I volunteered for watsonx.ai because I was familiar with it.
- I asked him for an example to model the work on and he recommended the Groq provider, which was extending both the OpenAI mixin and the LiteLLM mixin.
- So I re-implemented the watsonx.ai provider to extend both the OpenAI mixin and the LiteLLM mixin but it wouldn't work because the API is not OpenAI compatible.  Then I dropped the reference to the OpenAI mixin and now it works since LiteLLM does have explicit support for this API.
- Since I was reimplementing it anyway, I also changed it from using a static list of models to having it automatically pull the list of available models from the watsonx.ai server.  That makes it much more powerful because it brings in a lot of new models and also eliminates the old models that were on our static list but have been removed from watsonx.ai.  It also brings in all of the embedding models from watsonx.ai -- I don't believe any of those were on the static list before.

So what are the benefits here?  I think the original goal of standardizing on the mixins wound up being misleading because all of the others wound up being standardized on the OpenAI mixin and this one wound up being standardized on the LiteLLM mixin.  In theory there could be *some* benefit to being standardized on the LiteLLM mixin in the future if there is ever big demand for more providers that are compatible with LiteLLM but not the OpenAI client.  However, I think the main advantage of this PR is that it no longer uses a static list of supported models.

We certainly could revert to using the watsonx.ai client libraries instead of LiteLLM *and* also keep the change that adds dynamic model listing.  But then I think the question would be why do that work when this version is working now?

mattf commented:
> I believe there is an intention to only support openai compatible inference providers in a near future

there are no plans to exclude providers that don't implement an openai compatible api.

if we can simplify the WatsonX adapter by using LiteLLM, because LiteLLM has already done the hard adapting work, we should do it.

anyone who wants to exclude LiteLLM can clip the providers that use it.

pulling related utilities together and near to where they are used is a also a good outcome.

leseb commented:
> > I believe there is an intention to only support openai compatible inference providers in a near future
> 
> there are no plans to exclude providers that don't implement an openai compatible api.

There is plan but that's my intention, hope the team will align on that. Let's postpone that discussion though.

> 
> if we can simplify the WatsonX adapter by using LiteLLM, because LiteLLM has already done the hard adapting work, we should do it.
> 
> anyone who wants to exclude LiteLLM can clip the providers that use it.
> 
> pulling related utilities together and near to where they are used is a also a good outcome.

Alright, let's not waste any more time on this, the utilities for LiteLLM (the mixins) are good for now, let's use them. Let's revisit its removal later too. Thanks!
